File: OmniScript/py/OmniScript/omniscript/enginestatus.py
"""EngineStatus class.
"""
#Copyright (c) WildPackets, Inc. 2013-2014. All rights reserved.

import os
import sys
import logging

import config
import omniscript

from peektime import PeekTime


_engine_prop_dict = {
    'AdapterCount' : 'adapter_count',
    'AlarmCount' : 'alarm_count',
    'Captures' : 'capture_count',
    'CaptureSessionCount' : 'capture_session_count',
    'CpuCount' : 'cpu_count',
    'CpuType' : 'cpu_type',
    'DataFolder' : 'data_folder',
    'engine-type' : 'engine_type',
    'FileCount' : 'file_count',
    'FileVersion' : 'file_version',
    'FilterCount' : 'filter_count',
    'FiltersModTime' : 'filters_modification_time',
    'ForensicSearches' : 'forensic_search_count',
    'GraphCount' : 'graph_count',
    'LogTotalCount' : 'log_total_count',
    'MemAvailPhys' : 'memory_available_physical',
    'MemTotalPhys' : 'memory_total_physical',
    'Name' : 'name',
    'NameCount' : 'name_table_count',
    'NotificationCount' : 'notification_count',
    'OperatingSystem' : 'operating_system',
    'OS' : 'os',
    'Platform' : 'platform',
    'ProductVersion' : 'product_version',
    'StorageTotal' : 'storage_total',
    'StorageUsed' : 'storage_used',
    'Time' : 'time',
    'TimeZoneBias' : 'time_zone_bias',
    'Uptime' : 'uptime',
    }


def _parse_prop(a, v):
    result = None
    if a == 'storage_free_space':
        result = []
        for i in v.split(';'):
            k, v = i.strip(' ').split(' ', 1)
            # 69.4 GB -> 74,517,053,440 bytes
            cnt, unt, fs = v.split(' ', 2)
            cnt = float(cnt) * config.BYTES_IN_KILOBYTE # 1.23 -> 1230.0
            cnt = int(cnt)                              # -> 1230 kB
            if unt == 'MB':
                cnt = cnt * config.KILOBYTES_IN_MEGABYTE
            elif unt == 'GB':
                cnt = cnt * config.KILOBYTES_IN_GIGABYTE
            elif unt == 'TB':
                cnt = cnt * config.KILOBYTES_IN_TERABYTE
            result.append([k, cnt, fs])
    return result


class EngineStatus(object):
    """The Status of an :class:`OmniEngine <omniscript.omniengine.OmniEngine>`
    object.
    """

    adapter_count = 0
    """The number of adapters on the system."""

    alarm_count = 0
    """The number of alarms configured on the OmniEngine."""

    capture_count = 0
    """The number of Captures created on the OmniEngine."""

    capture_session_count = 0
    """The number of Capture Session on the OmniEngine."""

    cpu_count = 0
    """The number of CPUs on the system."""

    cpu_type = ''
    """The type of CPU on the system."""

    data_folder = ''
    """The default data folder of the OmniEngine."""

    engine_type = ''
    """The type of OmniEngine: OmniEngine Enterprise..."""

    file_count = 0
    """The number of Packet files in the OmniEngine's database."""

    file_version = ''
    """The File Version property of the OmniEngine."""

    filter_count = 0
    """The number of Filters on the OmniEngine."""

    filters_modification_time = None
    """The last time the Filters where modified."""

    forensic_search_count = 0
    """The number of Forensic Searches on the OmniEngine."""

    graph_count = 0
    """The number of Graphs on the OmniEngine."""

    host = ''
    """The address of the host system."""

    log_total_count = 0
    """The total number of Log entires on the OmniEngine."""

    memory_available_physical = 0
    """The amount of physical memory in bytes available on the system."""

    memory_total_physical = 0
    """The amount of physical memory in bytes on the system."""

    name = ''
    """The name of the OmniEngine."""

    name_table_count = 0
    """The number of Name Table entires (need to confirm) on the OmniEngine."""

    notification_count = 0
    """The number of Notifications on the OmniEngine."""

    operating_system = ''
    """The full name of the Operating System that the OmniEngine is
    running on.
    """

    os = ''
    """The short name of the Operating System that the OmniEngine is
    running on: Windows or Linux.
    """

    platform = 'localhost'
    """The platform of the system, Windows only: x64, win32."""

    port = 6367
    """The TCP Port of the connection to the host system."""

    product_version = ''
    """The Product Version property of the OmniEngine."""

    storage_available = 0
    """The amount of disk storage in bytes available on the system."""

    storage_total = 0
    """The total amount of disk storage in bytes on the system."""

    storage_used = 0
    """The amount of disk storage in bytes used on the system."""

    time = None
    """The current time, as
    :class:`PeekTime <omniscript.peektime.PeekTime>`,
    of day in UTC on the system."""

    time_zone_bias = 0
    """The time zone bias of the system."""

    uptime = 0
    """The number of nanoseconds the OmniEngine has been running."""

    def __init__(self, engine, logger, node=None):
        self._engine = engine
        self.logger = logger

        self.adapter_count = EngineStatus.adapter_count
        self.alarm_count = EngineStatus.alarm_count
        self.capture_count = EngineStatus.capture_count
        self.capture_session_count = EngineStatus.capture_session_count
        self.cpu_count = EngineStatus.cpu_count
        self.cpu_type = EngineStatus.cpu_type
        self.data_folder = EngineStatus.data_folder
        self.engine_type = EngineStatus.engine_type
        self.file_count = EngineStatus.file_count
        self.file_version = EngineStatus.file_version
        self.filter_count = EngineStatus.filter_count
        self.filters_modification_time = EngineStatus.filters_modification_time
        self.forensic_search_count = EngineStatus.forensic_search_count
        self.graph_count = EngineStatus.graph_count
        self.host = EngineStatus.host
        self.log_total_count = EngineStatus.log_total_count
        self.memory_available_physical = EngineStatus.memory_available_physical
        self.memory_total_physical = EngineStatus.memory_total_physical
        self.name = EngineStatus.name
        self.name_table_count = EngineStatus.name_table_count
        self.notification_count = EngineStatus.notification_count
        self.operating_system = EngineStatus.operating_system
        self.os = EngineStatus.os
        self.platform = EngineStatus.platform
        self.port = EngineStatus.port
        self.product_version = EngineStatus.product_version
        self.storage_available = EngineStatus.storage_available
        self.storage_total = EngineStatus.storage_total
        self.storage_used = EngineStatus.storage_used
        self.time = EngineStatus.time
        self.time_zone_bias = EngineStatus.time_zone_bias
        self.uptime = EngineStatus.uptime
        #Set provided values
        self.host = engine.host
        self.port = engine.port
        #Parse the dictionary.
        if node is not None:
            self._load(node)

    def __repr__(self):
        return 'EngineStatus: %s' % self.name

    def _load(self, node):
        """Set attributes from a dictionary."""
        for prop in node.findall('prop'):
            k = prop.get('name')
            v = prop.text
            a = _engine_prop_dict.get(k)
            if a is not None:
                if hasattr(self, a):
                    if isinstance(getattr(self, a), basestring):
                        setattr(self, a, v if v else '')
                    elif isinstance(getattr(self, a), (int, long)):
                        setattr(self, a, int(v) if v else 0)
                    elif isinstance(getattr(self, a), list):
                        setattr(self, a, _parse_prop(a, v))
                    if getattr(self, a) is None:
                        setattr(self, a, PeekTime(v))
            self.cpu_type = ' '.join(self.cpu_type.split())
            if self.storage_total > self.storage_used:
                self.storage_available = self.storage_total - self.storage_used

    def refresh(self):
        """Refresh the attributes of this object."""
        response = self._engine._issue_xml_command(config.OMNI_GET_STATUS, None)
        props = omniscript._parse_command_response(response, 'properties')
        if props is not None:
            self.filters_modification_time = EngineStatus.filters_modification_time
            self.time = EngineStatus.time
            self._load(props)

File: OmniScript/cs/OmniScript/EngineStatus.cs
// =============================================================================
// <copyright file="Adapter.cs" company="LiveAction, Inc.">
//  Copyright (c) 2018-2021 Savvius, Inc. All rights reserved.
//  Copyright (c) 2015-2018 Savvius, Inc. All rights reserved.
// </copyright>
// =============================================================================

namespace Savvius.Omni.OmniScript
{
    using System;
    using System.Runtime.InteropServices;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Xml.Linq;
    using System.Text;
    using System.Globalization;

    public class EngineStatus
    {
        public const String RootName = "properties";

        private OmniEngine Engine { get; set; }
        private OmniApi Connection { get; set; }
        public OmniLogger Logger { get; set; }

        /// <summary>
        /// Gets or sets the number of adapters.
        /// </summary>
        public uint AdapterCount { get; set; }

        /// <summary>
        /// Gets or sets the number of alarms.
        /// </summary>
        public uint AlarmCount { get; set; }

        /// <summary>
        /// Gets or sets the number of captures.
        /// </summary>
        public uint CaptureCount { get; set; }

        /// <summary>
        /// Gets or sets the number of CPUs on the host system.
        /// </summary>
        public uint CPUCount { get; set; }

        /// <summary>
        /// Gets or sets the type of CPU on the host system.
        /// </summary>
        public String CPUType { get; set; }

        /// <summary>
        /// Gets or sets the path of the Data Folder.
        /// </summary>
        public String DataFolder { get; set; }

        /// <summary>
        /// Gets or sets the type of Engine.
        /// </summary>
        public String EngineType { get; set; }

        /// <summary>
        /// Gets or sets the number of files.
        /// </summary>
        public uint FileCount { get; set; }

        /// <summary>
        /// Gets or sets the File Version of the Engine.
        /// </summary>
        public String FileVersion { get; set; }

        /// <summary>
        /// Gets or sets the number of filters.
        /// </summary>
        public uint FilterCount { get; set; }

        /// <summary>
        /// Gets or sets the last time the filters where modified.
        /// </summary>
        public PeekTime FilterModificationTime { get; set; }

        /// <summary>
        /// Gets or sets the number of Forensic Searches.
        /// </summary>
        public uint ForensicSearchCount { get; set; }

        /// <summary>
        /// Gets or sets the number of graphs.
        /// </summary>
        public uint GraphCount { get; set; }

        /// <summary>
        /// Gets or sets the name of the host system.
        /// </summary>
        public String Host { get; set; }

        /// <summary>
        /// Gets or sets the total number of log entries.
        /// </summary>
        public uint LogTotalCount { get; set; }

        /// <summary>
        /// Gets or sets the amount of physical memory available.
        /// </summary>
        public ulong MemoryAvailablePhysical { get; set; }

        /// <summary>
        /// Gets or sets the amount of total amount of physical memory.
        /// </summary>
        public ulong MemoryTotalPhysical { get; set; }

        /// <summary>
        /// Gets or sets the name of the Engine.
        /// </summary>
        public String Name { get; set; }

        /// <summary>
        /// Gets or sets the total number name table entries.
        /// </summary>
        public uint NameTableCount { get; set; }

        /// <summary>
        /// Gets or sets the number of notifications.
        /// </summary>
        public uint NotificationCount { get; set; }

        /// <summary>
        /// Gets or sets the name of the host system's operating system.
        /// </summary>
        public String OS { get; set; }

        /// <summary>
        /// Gets or sets the name of the host system's platform.
        /// </summary>
        public String Platform { get; set; }

        /// <summary>
        /// Gets or sets the communication port.
        /// </summary>
        public uint Port { get; set; }

        /// <summary>
        /// Gets or sets the Product version of the Engine.
        /// </summary>
        public String ProductVersion { get; set; }

        /// <summary>
        /// Gets or sets the storage space available.
        /// </summary>
        public ulong StorageAvailable { get; set; }

        /// <summary>
        /// Gets or sets the total amount of storage space.
        /// </summary>
        public ulong StorageTotal { get; set; }

        /// <summary>
        /// Gets or sets the amount of storage space used.
        /// </summary>
        public ulong StorageUsed { get; set; }

        /// <summary>
        /// Gets or sets the host system's time of day.
        /// </summary>
        public PeekTime Time { get; set; }

        /// <summary>
        /// Gets or sets the host system's timezone bias.
        /// </summary>
        public int TimeZoneBias { get; set; }

        /// <summary>
        /// Gets or sets the number of nanoseconds the Engine has been up.
        /// </summary>
        public PeekTime UpTime { get; set; }

        public EngineStatus(OmniEngine engine)
        {
            this.Engine = engine;
            this.Connection = engine.Connection;
            this.Logger = engine.Logger;

            this.AdapterCount = 0;
            this.AlarmCount = 0;
            this.CaptureCount = 0;
            this.CPUCount = 0;
            this.CPUType = "";
            this.DataFolder = "";
            this.EngineType = "";
            this.FileCount = 0;
            this.FileVersion = "";
            this.FilterCount = 0;
            this.FilterModificationTime = new PeekTime();
            this.ForensicSearchCount = 0;
            this.GraphCount = 0;
            this.Host = "";
            this.LogTotalCount = 0;
            this.MemoryAvailablePhysical = 0;
            this.MemoryTotalPhysical = 0;
            this.Name = "";
            this.NameTableCount = 0;
            this.NotificationCount = 0;
            this.OS = "";
            this.Platform = "";
            this.Port = 0;
            this.ProductVersion = "";
            this.StorageAvailable = 0;
            this.StorageTotal = 0;
            this.StorageUsed = 0;
            this.Time = new PeekTime();
            this.TimeZoneBias = 0;
            this.UpTime = new PeekTime();
        }

        public EngineStatus(OmniEngine engine, XElement status)
            : this(engine)
        {
            this.Load(status);
        }

        private void Load(XElement status)
        {
            if (status == null) return;

            IEnumerable<XElement> props = status.Elements("prop");
            foreach (XElement prop in props)
            {
                switch (prop.Attribute("name").Value)
                {
                    case "AdapterCount":
                        this.AdapterCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "AlarmCount":
                        this.AlarmCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "Captures":
                        this.CaptureCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "CaptureSessionCount":
                        this.ForensicSearchCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "CpuCount":
                        this.CPUCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "CpuType":
                        this.CPUType = prop.Value;
                        break;

                    case "DataFolder":
                        this.DataFolder = prop.Value;
                        break;

                    case "engine-type":
                        this.EngineType = prop.Value;
                        break;

                    case "FileCount":
                        this.FileCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "FileVersion":
                        this.FileVersion = prop.Value;
                        break;

                    case "FilterCount":
                        this.FilterCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "FiltersModTime":
                        this.FilterModificationTime = new PeekTime(prop.Value);
                        break;

                    case "ForensicSearches":
                        this.ForensicSearchCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "GraphCount":
                        this.GraphCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "LogTotalCount":
                        this.LogTotalCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "MemAvailPhys":
                        this.MemoryAvailablePhysical = Convert.ToUInt64(prop.Value);
                        break;

                    case "MemTotalPhys":
                        this.MemoryTotalPhysical = Convert.ToUInt64(prop.Value);
                        break;

                    case "Name":
                        this.Name = prop.Value;
                        break;

                    case "NameCount":
                        this.NameTableCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "NotificationCount":
                        this.NotificationCount = Convert.ToUInt32(prop.Value);
                        break;

                    case "OperatingSystem":
                        this.OS = prop.Value;
                        break;

                    case "OS":
                        this.OS = prop.Value;
                        break;

                    case "Platform":
                        this.Platform = prop.Value;
                        break;

                    case "ProductVersion":
                        this.ProductVersion = prop.Value;
                        break;

                    case "StorageTotal":
                        this.StorageTotal = Convert.ToUInt64(prop.Value);
                        break;

                    case "StorageUsed":
                        this.StorageUsed = Convert.ToUInt64(prop.Value);
                        break;

                    case "Time":
                        this.Time = new PeekTime(prop.Value);
                        break;

                    case "TimeZoneBias":
                        this.TimeZoneBias = Convert.ToInt32(prop.Value);
                        break;

                    case "Uptime":
                        this.UpTime = new PeekTime(prop.Value);
                        break;
                }
            }

            this.CPUType = String.Join(" ", this.CPUType.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries));
            this.StorageAvailable = this.StorageTotal = this.StorageUsed;
        }

        public void Refresh()
        {
            String response = this.Connection.XmlCommand( OmniApi.OMNI_GET_STATUS, null);
            XElement props = OmniApi.ParseResponse(response, EngineStatus.RootName);
            if (props != null)
            {
                this.Load(props);
            }
        }
    }
}
